ui/seek_bar.js

```diff
@@ -376,6 +465,155 @@ shaka.ui.SeekBar = class extends shaka.ui.RangeElement {
   updateAriaLabel_() {
     this.bar.ariaLabel = this.localization.resolve(shaka.ui.Locales.Ids.SEEK);
   }
+
+
+  /**
+   * @private
+   */
+  async showThumbnail_(pixelPosition, value) {
+    const thumbnailTrack = this.getThumbnailTrack_();
+    if (!thumbnailTrack) {
+      this.hideThumbnail_();
+      return;
+    }
+    if (value < 0) {
+      value = 0;
+    }
+    let thumbnail = this.thumbnails_[value];
+    const seekRange = this.player.seekRange();
+    if (this.player.isLive()) {
+      const totalSeconds = seekRange.end - value;
+      if (totalSeconds < 1) {
+        this.thumbnailTime_.textContent =
+            this.localization.resolve(shaka.ui.Locales.Ids.LIVE);
+      } else {
+        this.thumbnailTime_.textContent =
+            '-' + this.timeFormatter_(totalSeconds);
+      }
+    } else {
+      this.thumbnailTime_.textContent = this.timeFormatter_(value);
+    }
+    const offsetTop = -10;
+    const width = this.thumbnailContainer_.clientWidth;
+    let height = Math.floor(width * 9 / 16);
+    this.thumbnailContainer_.style.height = height + 'px';
+    this.thumbnailContainer_.style.top = -(height - offsetTop) + 'px';
+    const leftPosition = Math.min(this.bar.offsetWidth - width,
+        Math.max(0, pixelPosition - (width / 2)));
+    this.thumbnailContainer_.style.left = leftPosition + 'px';
+    this.thumbnailContainer_.style.visibility = 'visible';
+    if (!thumbnail) {
+      const playerValue = Math.max(Math.ceil(seekRange.start),
+          Math.min(Math.floor(seekRange.end), value));
+      thumbnail =
+          await this.player.getThumbnails(thumbnailTrack.id, playerValue);
+      this.thumbnails_[value] = thumbnail;
```

Comment:
The fact that thumbnails are cached based on value feels inefficient to me. We'll end up with multiple thumbnail objects for each image, if the player scrubs a lot...
I was casually playing around with the seek bar on one of the thumbnail tracks and ended up with nearly 600 objects in this map, each with a separate value for `imageWidth` and `imageHeight` and so on and so forth. And because the `getThumbnails` method is part of the public interface of Shaka Player, the variables won't be renamed in compiled mode!

Perhaps when you get a thumbnail this way, you could use the `startTime` and `duration` values to determine each `value` this thumbnail object is associated with, and then apply the thumbnail object to each of those values? I think that will result in a lower memory footprint because the `thumbnails_` dictionary will contain references and not copies, correct?

Reply:
You're right, it's pretty inefficient. For the moment I am going to eliminate the cache.